packages/metals-languageclient/src/fetchMetals.ts

```diff
@@ -1,56 +1,114 @@
     (p) => !p.startsWith("-agentlib")
   );
-
   const serverDependency = calcServerDependency(serverVersion);
-  return spawn(
-    javaPath,
-    [
-      ...javaOptions,
-      ...fetchProperties,
-      "-Dfile.encoding=UTF-8",
-      "-jar",
-      coursierPath,
-      "fetch",
-      "-p",
-      "--ttl",
-      // Use infinite ttl to avoid redunant "Checking..." logs when using SNAPSHOT
-      // versions. Metals SNAPSHOT releases are effectively immutable since we
-      // never publish the same version twice.
-      "Inf",
-      serverDependency,
-      "-r",
-      "bintray:scalacenter/releases",
-      "-r",
-      "sonatype:public",
-      "-r",
-      "sonatype:snapshots",
-      "-p",
-    ],
-    {
-      env: {
-        COURSIER_NO_TERM: "true",
-        ...extraEnv,
-        ...process.env,
-      },
-      stdio: ["ignore"], // Due to Issue: #219
+
+  const coursierArgs = [
+    "fetch",
+    "-p",
+    "--ttl",
+    // Use infinite ttl to avoid redunant "Checking..." logs when using SNAPSHOT
+    // versions. Metals SNAPSHOT releases are effectively immutable since we
+    // never publish the same version twice.
+    "Inf",
+    serverDependency,
+    "-r",
+    "bintray:scalacenter/releases",
+    "-r",
+    "sonatype:public",
+    "-r",
+    "sonatype:snapshots",
+    "-p",
+  ];
+
+  const path = process.env["PATH"];
+  let possibleCoursier: string | undefined;
+  if (path) {
+    possibleCoursier = validateCoursier(path);
+  }
+
+  function spawnDefault(): ChildProcessPromise {
+    return spawn(
+      javaPath,
+      [
+        ...javaOptions,
+        ...fetchProperties,
+        "-Dfile.encoding=UTF-8",
+        "-jar",
+        coursierPath,
+      ].concat(coursierArgs),
+      {
+        env: {
+          COURSIER_NO_TERM: "true",
+          ...extraEnv,
+          ...process.env,
+        },
+      }
+    );
+  }
+
+  if (possibleCoursier) {
+    const coursier: string = possibleCoursier;
+    const coursierVersion = spawnSync(coursier, ["version"]);
+    if (coursierVersion.status !== 0) {
+      return spawnDefault();
+    } else {
+      output.appendLine(`Using coursier located at ${coursier}`);
+      return spawn(coursier, coursierArgs);
     }
-  );
+  } else {
+    return spawnDefault();
+  }
+}
+
+export function validateCoursier(pathEnv: string): string | undefined {
+  const possibleCoursier = pathEnv
+    .split(path.delimiter)
+    .flatMap((p) => {
+      try {
+        if (fs.statSync(p).isDirectory()) {
+          return fs.readdirSync(p).map((sub) => path.resolve(p, sub));
+        } else return [p];
+      } catch (e) {
+        return [];
+      }
+    })
+    .find(
+      (p) =>
+        p.endsWith(path.sep + "cs") ||
+        p.endsWith(path.sep + "coursier") ||
+        (process.platform === "win32" && p.endsWith(path.sep + "cs.bat"))
+    );
+  if (possibleCoursier) {
+    const coursierVersion = spawnSync(possibleCoursier, ["version"]);
```

Comment:
Not sure how to work around this, I can't find a way to chain ChildProcess since then is only available on Promise. Alternatively, we could make a spawn with a wait to make sure this doesn't take too long. Ideas?

Reply:
I think I did? Doing `.then` on ChildProcessPromise was only returning `Promise<Output>` and not sure of there is a way around it really.

Reply:
Maybe https://github.com/sindresorhus/execa help?
As long as we use raw child_process stuffs (without writing a wrapper) there's no way to do that (get stdout wrapped by Promise if success, and Promise failed if return code isn't 0).

Reply:
Doesn't look like it :/ We want to pipe the output from coursier to VS Code output, but it seems I can do it only via ChildProcessPromise. I tried returning a Promise<ChildProcessPromise>, but that seems to have a wrong type in the actual javascript underneath:
https://github.com/scalameta/metals-vscode/commit/906323d33891b9318e95266618627b4445e85cdd

Reply:
Och actually if I have an intermediate object it works, not 100% pretty, but feels ok.